course_catalog/etl/youtube.py

```diff
@@ -424,3 +428,76 @@ def transform(extracted_channels):
                 for playlist, videos in playlists
             ),
         }
+
+
+def get_youtube_videos_for_transcripts_job(
+    *, created_after, created_minutes, overwrite=False
+):
+    """
+    course_catalog.Video object filtered to tasks.get_youtube_transcripts job params
+
+    Args:
+        created_after (date or None):
+            if a date inclued only videos with a created_on after that date
+        created_minutes (int or None):
+            if an int include only videos with created_on in the last created_minutes minutes
+        overwrite (bool):
+            if true include videos that already have transcripts
+
+    Returns
+        Django filtered course_catalog.videos object
+    """
+
+    videos = Video.objects.all()
+    if not overwrite:
+        videos = videos.filter(transcript="", published=True)
+
+    if created_after:
+        videos = videos.filter(created_on__gte=created_after)
+    elif created_minutes:
+        date = now_in_utc() - timedelta(minutes=created_minutes)
+        videos = videos.filter(created_on__gte=date)
+    return videos
+
+
+def get_youtube_transcripts(videos):
+    """
+    Fetch transcripts for Youtube videos
+
+    Args:
+        vidoes - collection of course_catalog.Video objects
+    """
+
+    # The call to download the transcript occasionally fails. We'll retry once after the first failure.
+    # If 15 consecutive  videos fail to load with pytube.exceptions.VideoUnavailable error
+    # we will assume we are being rate limited and stop the job early
+
+    consecutive_video_unavailable_failures = 0
+    max_consecutive_video_unavailable_failures = 15
+    for video in videos:
+        if (
+            consecutive_video_unavailable_failures
+            >= max_consecutive_video_unavailable_failures
+        ):
+            log.error(
+                "%i consecutive faliures for transcript downloads. Ending transcript download job early. ",
+                max_consecutive_video_unavailable_failures,
+            )
+            break
+
+        tries = 2
+        for attempt in range(tries):
+            try:
+                caption = get_captions_for_video(video)
+            except (pytube.exceptions.PytubeError, KeyError) as error:
+                if attempt == tries - 1:
+                    log.error("Unable to fetch transcript for video id=%i", video.id)
+                    if isinstance(error, pytube.exceptions.VideoUnavailable):
```

Comment:
I updated the code to only give up after 15 pytube.exceptions.VideoUnavailable errors, since that is the error type for rate limiting. pytube.exceptions.VideoUnavailable is a child of pytube.exceptions.PytubeError